mujoco_warp/_src/io.py

```diff
@@ -1303,6 +1304,547 @@ def reset_contact(
   )
 
 
+def set_const_fixed(m: types.Model, d: types.Data):
+  """Compute fixed quantities (independent of qpos0).
+
+  Computes:
+    - body_subtreemass: mass of body and all descendants (depends on body_mass)
+    - ngravcomp: count of bodies with gravity compensation (depends on body_gravcomp)
+
+  Args:
+    m: The model containing kinematic and dynamic information (device).
+    d: The data object containing the current state and output arrays (device).
+  """
+
+  @nested_kernel(module="unique", enable_backward=False)
+  def init_subtreemass(
+    body_mass_in: wp.array2d(dtype=float),
+    body_subtreemass_out: wp.array2d(dtype=float),
+  ):
+    worldid, bodyid = wp.tid()
+    body_mass_id = worldid % body_mass_in.shape[0]
+    body_subtreemass_id = worldid % body_subtreemass_out.shape[0]
+    body_subtreemass_out[body_subtreemass_id, bodyid] = body_mass_in[body_mass_id, bodyid]
+
+  @nested_kernel(module="unique", enable_backward=False)
+  def accumulate_subtreemass(
+    bodyid: int,
+    body_parentid: wp.array(dtype=int),
+    body_subtreemass_io: wp.array2d(dtype=float),
+  ):
+    worldid = wp.tid()
+    body_subtreemass_id = worldid % body_subtreemass_io.shape[0]
+    parentid = body_parentid[bodyid]
+    body_subtreemass_io[body_subtreemass_id, parentid] += body_subtreemass_io[body_subtreemass_id, bodyid]
+
+  wp.launch(init_subtreemass, dim=(d.nworld, m.nbody), inputs=[m.body_mass], outputs=[m.body_subtreemass])
+  for bodyid in range(m.nbody - 1, 0, -1):
```

Comment:
instead of iterating over bodies serially i think we can leverage tree-level parallelism. see [com_pos](https://github.com/google-deepmind/mujoco_warp/blob/6b7650ec5a92680d2ed3ad390657ca173bc26ec6/mujoco_warp/_src/smooth.py#L489) for a design pattern that should be relevant here for reverse tree traversal with tree-level parallelism.